### ISSUE TITLE
[ET-VK][Adreno] Low-rate non-deterministic CNN output, dependent on GPU power state; clean on Mali

### ISSUE THREAD

**Issue by msluszniak** [pytorch/executorch#21939]
### 🐛 Describe the bug

On Adreno, CNNs lowered to the Vulkan delegate produce different results run to run, at a low but persistent rate, even with the im2col/GEMM path from #21938 disabled. EfficientNetV2-S gives **24 distinct results in 25 replays of the same already-encoded command buffer inside a single process**, which rules out load-time state, prepack and weight packing.

Deviation magnitude: EfficientNetV2-S logits span [-1.44, 2.75]; typical deviations are 0.001 to 0.009, with rare excursions to about 4.0. argmax was stable across 30 runs for one input, so class flips need the rare large excursions.

### This is very likely an Adreno driver or hardware problem, not an ExecuTorch bug

1. **A different GPU vendor is clean.** Same binaries, same `.pte`, Galaxy S10+ (Exynos 9820, **Mali-G76**): 20 runs, 4020 in-process replays, **zero deviations**. The probe below that deviates on Adreno gives 0 across 1200 replays there. Output is also numerically correct on Mali, not merely stable: matched Vulkan and CPU exports from one model instance both land within fp32 rounding of the eager reference (9.9e-07 and 3.6e-07).
Confirmed at the application level as well, not just with `executor_runner`: a demo app built from the stock Vulkan backend runs both object detection (YOLO26) and image classification (EfficientNetV2-S) on the S10 with correct and stable results.

Separately, and worth knowing for anyone weighing the Vulkan delegate on Mali: on this S10 the Vulkan delegate is close to **2x slower** than the XNNPACK equivalent for the same models. That is the opposite of the Adreno device, where Vulkan is clearly faster. The S10 is a 2019 part so this is not necessarily representative of current Mali hardware, but it does mean the delegate is fast on the GPU where it is incorrect and slow on the GPU where it is correct.

2. **Khronos synchronization validation reports zero hazards** on the failing workload. Layer injected into a debuggable app via `gpu_debug_layers`, sync validation confirmed active by the layer's own settings message. No `SYNC-HAZARD`, no VUID errors of any kind.
3. **Unconditional full memory barriers change nothing.** Patching `CommandBuffer::insert_barrier` to always emit a global `ALL_COMMANDS -> ALL_COMMANDS` barrier with `VK_ACCESS_MEMORY_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT` before every dispatch, on top of the computed per-resource barriers, does not help.

The dispatch sequence is also deterministic: an env-gated trace in `DispatchNode::encode` gives byte-identical traces across runs.

### Measurement note, important for anyone reproducing

The failure rate is strongly dependent on GPU power state and swings by more than an order of magnitude for the same `.pte` and same binary within minutes. Same build, EfficientNetV2-S: 79 distinct / 101 at ~1000 MHz, 13 / 101 at ~826 MHz, 1 / 101 pinned at 461 MHz.

`adb shell cmd power set-fixed-performance-mode-enabled true` pins the GPU at 461 MHz without root on this device (`devfreq/{min,max}_freq` are unavailable). Over 4 alternating pairs of 100 in-process replays: 3 excess distinct / 404 pinned versus 30 / 404 unpinned.

Consequences for methodology:
- Any A/B comparison that is not interleaved within a single round is unreliable.
- Patch `executor_runner` to dump outputs on every iteration of `--num_executions`; stock writes only the last. This is far more sensitive than relaunching the process.
- Probe models must be activation calibrated. Untrained stacked convs collapse the signal to ~1e-19, where "distinct output" counts last-bit noise on values carrying no information.

### Smallest structure that shows it

12 rounds, order rotated each round, 200 in-process replays per run, calibrated probes, input `(1, 48, 64, 64)`, im2col disabled. Excess distinct outputs per 2400 replays:

| graph | run A | run B |
|---|--:|--:|
| conv3x3 -> SiLU -> conv1x1 | 43 | 18 |
| conv3x3 -> ReLU -> conv1x1 | 3 | 11 |
| conv3x3 -> Sigmoid -> conv1x1 | - | 2 |
| conv3x3 -> conv1x1 | 1 | - |
| conv3x3 -> SiLU -> conv3x3 | 0 | 0 |

`conv -> SiLU -> conv3x3` is 0 across 4800 replays while the same graph with a 1x1 consumer is the worst case. With no activation at all, kernel size alone does nothing (0 to 6 per 2400 for all four combinations of 1x1/3x3 producer and consumer). Absolute rates are low, so treat this as a lead rather than a conclusion. It is only visible with DVFS active; pinned at 461 MHz all four collapse to the noise floor.

Forcing 1x1 convs away from `Conv2dMethod::Pointwise` into the generic sliding window shader does **not** fix it, so it is not shader identity. The same generic shader is clean at 3x3 and dirty at 1x1.

### Ruled out, each tested

Memory aliasing (`add_tensor(..., -1)` instead of `mem_obj_id()`; made it worse), storage type, command buffer splitting (`execute_threshold_node_count` set to `SIZE_MAX`), descriptor pool recycling (`descriptor_pool_safety_factor` 1.25 to 64.0), caller threading, host cache coherency, node count.

Possibly the same underlying defect as #21938, which is also Adreno-only but is 100% reproducible and independent of clock.

### Versions

ExecuTorch 1.4.1. Failing: Galaxy S26 Ultra (SM-S948B), Adreno 840v2. Clean: Galaxy S10+ (SM-G975F), Mali-G76.

**Comment by msluszniak** [comment]
Update: I have a **much smaller, parameterised reproducer** for this issue, and four more causes ruled out.

### Synthetic reproducer (no torchvision needed)

Repeating this block `n` times, with the im2col path disabled so #21938 is out of the picture:

```python
torch.nn.Conv2d(48, 192, 3, padding=1, bias=False)
torch.nn.SiLU()
torch.nn.Conv2d(192, 48, 1, bias=False)
```

input `(1, 48, 64, 64)`, 40 runs of `executor_runner`, counting distinct md5 of stdout:

| n | dispatches | distinct / 40 |
|--:|-----------:|--------------:|
| 8  | 34 | 2 |
| 9  | 38 | 5 |
| 10 | 42 | 7 |
| 11 | 46 | 2 |
| 12 | 50 | 7 |
| 24 | 98 | 19 / 30 |

So this is **probabilistic and scales with the number of dispatches** rather than having a clean threshold — with enough repeats even the 34-dispatch case fails. That matches the observed model behaviour: EfficientNetV2-S (hundreds of dispatches) is wrong essentially every run, while shallower graphs look fine until you run them enough times.

The **dispatch sequence itself is deterministic** — dumping `ETVK_DISPATCH_TRACE` for two runs of the same model gives byte-identical traces (same kernels, same order, same read/write ValueRef sets), so the variation is in execution, not in encoding or planning.

### Additionally ruled out

On top of the list in the original post, none of these fix it (each tested with the im2col path forced off, 30 runs per model, on EfficientNetV2-S, YOLO26-nano and the chains above):

- **Command-buffer splitting** — `execute_threshold_node_count` / `execute_initial_threshold_node_count` set to `SIZE_MAX`, so the whole graph encodes into a single command buffer with no intermediate unsynchronised submits. Still fails (chain12 7/30, EfficientNetV2-S 20/30, YOLO26-nano 28/30).
- **Descriptor pool exhaustion / set recycling** — `descriptor_pool_safety_factor` raised from 1.25 to 64.0. Still fails (chain12 6/30, EfficientNetV2-S 21/30, YOLO26-nano 24/30).
- **Memory aliasing** — `add_tensor(..., -1)` instead of `tensor_fb->mem_obj_id()`. Still fails.
- **Kernel mix** — the failing and passing slices use the same set of shaders (`conv2d_float`, `conv2d_pw_tiled_float`, `sigmoid_float_texture3d`, `binary_mul_texture3d_float`, `binary_add_texture3d_float`); the failing one simply has more of them.

Given the rate scales with dispatch count and the encoded sequence is stable, the remaining suspect I would look at is the **pipeline barriers inserted between dependent dispatches inside a single command buffer** (built from the `ArgGroup` access flags in `bind_values_to_descriptor_set`). If any op's declared access does not capture a real read-after-write dependency, the resulting missing barrier would behave exactly like this. I have not been able to confirm that from source alone.

Happy to run further experiments — the loop above takes seconds per iteration on the device.

**Comment by msluszniak** [comment]
**Correction to my previous comment, plus new data.** Two things in that update were wrong: I had not controlled for GPU power state, and my synthetic probes were numerically degenerate.

### 1. The synthetic reproducer table above should be disregarded

Those chained `Conv2d -> SiLU -> Conv2d` probes produce outputs around 1e-19 after 24 blocks, because untrained stacked convs collapse the signal. "Distinct md5 of stdout" was therefore counting last-bit noise on values carrying no information. Rebuilt with per-layer calibration (each conv weight divided by `layer(h).std()`, so activations stay at std=1), the depth scaling does not reproduce.

### 2. Missing pipeline barriers are ruled out

I ended that comment suggesting a missing barrier between dependent dispatches. It is not that. Patching `CommandBuffer::insert_barrier` to unconditionally emit a global `ALL_COMMANDS -> ALL_COMMANDS` barrier with `VK_ACCESS_MEMORY_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT` before every dispatch, on top of the computed per-resource barriers, changes nothing:

| model | stock | full barriers |
|---|--:|--:|
| chain12 | 6 distinct / 30 | 5 / 30 |
| chain24 | 16 / 30 | 22 / 30 |

### Measurement note for anyone reproducing this

The failure rate is dominated by GPU power state and swings by over an order of magnitude for the same .pte and same binary within minutes. Same build, EfficientNetV2-S: 79 distinct / 101 at ~1000 MHz, 13 / 101 at ~826 MHz, 1 / 101 pinned at 461 MHz.

On this device `adb shell cmd power set-fixed-performance-mode-enabled true` pins the GPU at 461 MHz without root (`devfreq/{min,max}_freq` are not available). Over 4 alternating pairs of 100 in-process replays: 3 excess distinct / 404 pinned versus 30 / 404 unpinned.

So any A/B not interleaved within a single round is unreliable. The numbers in the issue body deserve the same scepticism.

### A better harness

Patching `executor_runner` to dump outputs on every iteration of `--num_executions` (stock writes only the last) is much more sensitive, and separates load-time state from execution. EfficientNetV2-S gives **24 distinct results in 25 replays of the same already-encoded command buffer inside one process**. That rules out prepack, weight packing, and any load-time state. The dispatch sequence is deterministic too: `ETVK_DISPATCH_TRACE` yields byte-identical traces across runs.

### Deviation magnitude

EfficientNetV2-S logits span [-1.44, 2.75]. Typical deviations are 0.001 to 0.009, with rare excursions to about 4.0. argmax was stable across 30 runs for that input, so class flips need the rare large excursions.

### What does reproduce under a controlled design

12 rounds, order rotated each round so position cannot be confounded with model, 200 in-process replays per run, all probes calibrated, input `(1, 48, 64, 64)`. Counting distinct outputs beyond the first:

| graph | excess distinct / 2400 | mean clk |
|---|--:|--:|
| conv3x3 -> SiLU -> conv1x1 | 43 | 575 MHz |
| conv3x3 -> ReLU -> conv1x1 | 3 | 672 MHz |
| conv3x3 -> conv1x1 | 1 | 579 MHz |
| conv3x3 -> SiLU -> conv3x3 | 0 | 588 MHz |

Second campaign, same design, varying the activation:

| graph | excess distinct / 2400 | mean clk |
|---|--:|--:|
| conv3x3 -> SiLU -> conv1x1 | 18 | 776 MHz |
| conv3x3 -> ReLU -> conv1x1 | 11 | 771 MHz |
| conv3x3 -> Sigmoid -> conv1x1 | 2 | 762 MHz |
| conv3x3 -> SiLU -> conv3x3 | 0 | 767 MHz |

With no activation at all, a 2x2 over kernel sizes:

| graph | excess distinct / 2400 |
|---|--:|
| conv3x3 -> conv1x1 | 0 |
| conv3x3 -> conv3x3 | 0 |
| conv1x1 -> conv1x1 | 6 |
| conv1x1 -> conv3x3 | 3 |

The consistent result across both campaigns is `conv -> SiLU -> conv3x3` at 0 / 4800 replays, while the same graph with a 1x1 consumer is the worst case. Absolute rates are low, so this is a lead rather than a conclusion. Note it is only visible with DVFS active: pinned at 461 MHz all four collapse to the noise floor (0 to 9 per 2400).

### It is not the pointwise shader itself

Since a 1x1 consumer looked necessary, I forced 1x1 convs away from `Conv2dMethod::Pointwise` into the generic sliding window `conv2d` shader. That does not fix it. Partial A/B, interrupted at 15 runs of 200 replays each:

| | stock | 1x1 forced to sliding window |
|---|--:|--:|
| EfficientNetV2-S | 153 excess over 3 runs | 203 over 4 runs |
| conv3x3 -> SiLU -> conv1x1 | 11 over 4 runs | 136 over 4 runs |

So it is not shader identity. The same generic shader is clean at 3x3 and dirty at 1x1, which points at something tied to the 1x1 configuration (workgroup sizing, loop trip counts, channel packing) rather than the pointwise implementation.

### Also ruled out, each tested

Memory aliasing (`add_tensor(..., -1)` instead of `mem_obj_id()`; made it worse, not better), storage type, command buffer splitting (`execute_threshold_node_count` set to `SIZE_MAX`), descriptor pool recycling (`descriptor_pool_safety_factor` 1.25 to 64.0), caller threading, host cache coherency.

Device: Galaxy S26 Ultra (SM-S948B), Adreno 840v2, ExecuTorch 1.4.1.

**Comment by SS-JIA** [comment]
Currently investigating; I will post updates in https://github.com/pytorch/executorch/issues/21938 for now.
